primitives/maybe-compressed-blob/src/lib.rs

```
@@ -0,0 +1,157 @@
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+		match *self {
+			Error::PossibleBomb => write!(f, "Possible compression bomb encountered"),
+			Error::Invalid => write!(f, "Blob had invalid format"),
+		}
+	}
+}
+
+impl std::error::Error for Error { }
+
+fn read_from_decoder(
+	decoder: impl Read,
+	blob_len: usize,
+	bomb_limit: usize,
+) -> Result<Vec<u8>, Error> {
+	let mut decoder = decoder.take((bomb_limit + 1) as u64);
+
+	let mut buf = Vec::with_capacity(blob_len);
+	decoder.read_to_end(&mut buf).map_err(|_| Error::Invalid)?;
+
+	if buf.len() <= bomb_limit {
+		Ok(buf)
+	} else {
+		Err(Error::PossibleBomb)
+	}
+}
+
+#[cfg(not(target_os = "unknown"))]
+fn decompress_zstd(blob: &[u8], bomb_limit: usize) -> Result<Vec<u8>, Error> {
+	let decoder = zstd::Decoder::new(blob).map_err(|_| Error::Invalid)?;
+
+	read_from_decoder(decoder, blob.len(), bomb_limit)
+}
+
+#[cfg(target_os = "unknown")]
+fn decompress_zstd(mut blob: &[u8], bomb_limit: usize) -> Result<Vec<u8>, Error> {
+	let blob_len = blob.len();
+	let decoder = ruzstd::streaming_decoder::StreamingDecoder::new(&mut blob)
+		.map_err(|_| Error::Invalid)?;
+
+	read_from_decoder(decoder, blob_len, bomb_limit)
+}
+
+/// Decode a blob, if it indicates that it is compressed. Provide a `bomb_limit`, which
+/// is the limit of bytes which should be decompressed from the blob.
+pub fn decompress(blob: &[u8], bomb_limit: usize) -> Result<Cow<[u8]>, Error> {
+	if blob.starts_with(&ZSTD_PREFIX) {
+		decompress_zstd(&blob[ZSTD_PREFIX.len()..], bomb_limit).map(Into::into)
+	} else {
+		Ok(blob.into())
+	}
+}
+
+/// Encode a blob as compressed. If the blob's size is over the bomb limit,
+/// this will not compress the blob, as the decoder will not be able to be
+/// able to differentiate it from a compression bomb.
+#[cfg(not(target_os = "unknown"))]
+pub fn compress(blob: &[u8], bomb_limit: usize) -> Option<Vec<u8>> {
+	use std::io::Write;
+
+	if blob.len() > bomb_limit {
+		return None;
```

I think we should return an error here. Otherwise we could drop `bomb_limit` directly.

---

Because the semantic of an error is different? The function is documented as returning `None` if the size exceeds the bomb limit.